test/extensions/filters/http/jwt_authn/authenticator_test.cc

```diff
@@ -174,8 +174,60 @@ TEST_F(AuthenticatorTest, TestJwtWithNonExistKid) {
   expectVerifyStatus(Status::JwtVerificationFail, headers);
 }
 
-// This test verifies the Jwt without "iss" work
-TEST_F(AuthenticatorTest, TestJwtWithoutIss) {
+// Jwt "iss" is "other.com", it doesn't match "issuer" specified in JwtProvider.
+// The verification fails with JwtUnknownIssuer.
+TEST_F(AuthenticatorTest, TestWrongIssuer) {
+  EXPECT_CALL(*raw_fetcher_, fetch(_, _, _)).Times(0);
+  // Token with other issuer should fail.
+  Http::TestRequestHeaderMapImpl headers{
+      {"Authorization", "Bearer " + std::string(OtherGoodToken)}};
+  expectVerifyStatus(Status::JwtUnknownIssuer, headers);
+}
+
+// Jwt "iss" is "other.com", "issuer" in JwtProvider is not specified,
+// authenticator has a valid provider. The verification is OK.
+TEST_F(AuthenticatorTest, TestWrongIssuerOKWithProvider) {
+  auto& provider = (*proto_config_.mutable_providers())[std::string(ProviderName)];
+  provider.clear_issuer();
+  provider.clear_audiences();
+  createAuthenticator();
+
+  EXPECT_CALL(*raw_fetcher_, fetch(_, _, _))
+      .WillOnce(Invoke([this](const envoy::config::core::v3::HttpUri&, Tracing::Span&,
+                              JwksFetcher::JwksReceiver& receiver) {
+        receiver.onJwksSuccess(std::move(jwks_));
+      }));
+
+  // Token with other issuer is OK since "issuer" is not specified in the provider.
+  Http::TestRequestHeaderMapImpl headers{
+      {"Authorization", "Bearer " + std::string(OtherGoodToken)}};
+  expectVerifyStatus(Status::Ok, headers);
+}
+
+// Jwt "iss" is "other.com", "issuer" in JwtProvider is not specified,
+// authenticator doesn't have a valid provider. The verification is OK.
```

Comment:
why is the verification ok when the authenticator doesn't have a valid provider? shouldn't it fail in this case.

Also in what case will the authenticator not have a valid provider?

Reply:
I added comment in source/extensions/filters/http/jwt_authn/authenticator.cc line 163 when provider is not valid.
I will copy it here to readability.

Reply:
If a Provider doesn't have "issuer" specified,  it can be used for any Jwt in the invalid provider case, which is in "allow_missing" or "allow_failed" case.